File: src/lib/php/DSTemplate.php
<?php

require_once "DSLib.php";
require_once "DSClassInfo.php";
require_once "DSConstants.php";

define("DS_TEMPLATE_VERSION", "0.0.1");
define("DS_TEMPLATE_NAME",    "DSTemplate");
define("DS_TEMPLATE_AUTHOR",  "Thorsten Geppert (thorsten@ext2fs.de)");

class DSTemplate extends DSLib {

	protected $root          = "";
	protected $content       = "";
	protected $parsedContent = "";
	protected $filename      = "";
	protected $intern        = array();
	protected $dynamic       = array();
	protected $variables     = array();

	public function __construct($root = "") {
		parent::__construct();

		$this->root = $root;
		if(strlen($this->root) > 0)
			if($this->root[strlen($this->root) - 1] != DS_PATH_SEPARATOR)
				$this->root .= "/";
		$this->FillInternVariables();
	}

	public function GetClassInfo() {
		return new DSClassInfo(
			DS_TEMPLATE_NAME,
			DS_TEMPLATE_VERSION,
			DS_TEMPLATE_AUTHOR
		);
	}

	protected function FillInternVariables() {
		$this->SetInternVariable("date", date("Y-m-d"));
		$this->SetInternVariable("time", date("h:i:s"));
		$this->SetInternVariable("datetime", date("Y-m-d"));
	}

	public function GetRoot() {
		return $this->root;
	}

	public function SetContent($content) {
		$this->content = $content;
	}

	public function SetFile($filename) {
		$this->filename = $filename;
		if($this->content = @file_get_contents($this->root.$filename))
			return DS_FILE_NOT_READ;
		return DS_OK;
	}

	public function &GetContent() {
		return $this->content;
	}

	public function GetContentCopy() {
		return $this->content;
	}

	public function &Get() {
		return $this->parsedContent;
	}

	public function GetCopy() {
		return $this->parsedContent;
	}

	public function SetVariable(
		$key,
		$value,
		$htmlSpecialChars = false
	) {
		$this->variables[$key] = $htmlSpecialChars ? htmlspecialchars($value) : $value;	
	}

	public function GetVariable($key) {
		return $this->variables[$key];
	}

	public function SetInternVariable(
		$key,
		$value,
		$htmlSpecialChars = false
	) {
		$this->intern[$key] = $htmlSpecialChars ? htmlspecialchars($value) : $value;
	}

	public function GetInternVariable($key) {
		return $this->intern[$key];
	}

	public function GetInternVariables() {
		return $this->intern;
	}

	public function SetDynamicVariable(
		$key,
		$value,
		$htmlSpecialChars = false
	) {
		$this->dynamic[$key] = $htmlSpecialChars ? htmlspecialchars($value) : $value;
	}

	public function GetDynamicVariable($key) {
		return $this->dynamic[$key];
	}

	public function Parse() {
		$tmp = $this->content;

		foreach($this->variables as $key => $value)
			$tmp = str_replace("{s:".$key."}", $value, $tmp);
		foreach($this->intern as $key => $value)
			$tmp = str_replace("{i:".$key."}", $value, $tmp);
		foreach($this->dynamic as $key => $value)
			$tmp = str_replace("{d:".$key."}", $value, $tmp);

		$this->parsedContent .= $tmp;

		return $tmp;
	}

	public function Show() {
		echo $this->parsedContent;
	}

	public function ClearParsed() {
		$this->parsedContent = array();
	}

}

?>

File: src/lib/cpp/DSTemplate.cpp
#include "DSTemplate.h"
#include "DSConstants.h"
#include "DSCasts.h"
#include "DSDateTime.h"
#include <wx/ffile.h>
#if ! defined __WXMSW__ && ! defined __WXMAC__
	#include "DSCgi.h"
	#include <fcgi_stdio.h>
#endif

DynSoft::DSTemplate::DSTemplate(const wxString &root) : DSLib() {
	this->root = root;
	if(this->root.Length() > 0)
		if(this->root[this->root.Length() - 1] != DS_PATH_SEPARATOR)
			this->root.Append(DS_PATH_SEPARATOR);
	FillInternVariables();
}

void DynSoft::DSTemplate::FillInternVariables() {
	DSDateTime dt;
	SetInternVariable(_N("date"), dt.ToDatabaseDate());
	SetInternVariable(_N("time"), dt.ToDatabaseTime());
	SetInternVariable(_N("datetime"), dt.ToDatabaseDateTime());
}

wxString DynSoft::DSTemplate::GetRoot() const {
	return root;
}

void DynSoft::DSTemplate::SetContent(const wxString &content) {
	this->content = content;
}

int DynSoft::DSTemplate::SetFile(const wxString &filename) {
	this->filename = filename;
	wxFFile file(root + filename, _N("r"));
	if(!file.IsOpened())
		return DS_FILE_NOT_OPEN;
	
	int result = DS_OK;
	if(!file.ReadAll(&content))
		result = DS_FILE_NOT_READ;
	file.Close();

	return result;
}

wxString &DynSoft::DSTemplate::GetContent() {
	return content;
}

wxString DynSoft::DSTemplate::GetContentCopy() const {
	return content;
}

wxString DynSoft::DSTemplate::GetFilename() const {
	return filename;
}

wxString &DynSoft::DSTemplate::Get() {
	return parsedContent;
}

wxString DynSoft::DSTemplate::GetCopy() const {
	return parsedContent;
}

void DynSoft::DSTemplate::SetVariable(
	const wxString &key,
	const wxString &value,
	const bool htmlSpecialChars
) {
	#if ! defined __WXMSW__ && ! defined __WXMAC__
		variables[key] = htmlSpecialChars ? DSCgi::HtmlSpecialChars(value) : value;
	#else
		variables[key] = value;
	#endif
}

wxString DynSoft::DSTemplate::GetVariable(const wxString &key) {
	return variables[key];
}

void DynSoft::DSTemplate::SetInternVariable(
	const wxString &key,
	const wxString &value,
	const bool htmlSpecialChars
) {
	#if ! defined __WXMSW__ && ! defined __WXMAC__
		intern[key] = htmlSpecialChars ? DSCgi::HtmlSpecialChars(value) : value;
	#else
		intern[key] = value;
	#endif
}

wxString DynSoft::DSTemplate::GetInternVariable(const wxString &key) {
	return intern[key];
}

DSStringHashMap DynSoft::DSTemplate::GetInternVariables() const {
	return intern;
}

void DynSoft::DSTemplate::SetDynamicVariable(
	const wxString &key,
	const wxString &value,
	const bool htmlSpecialChars
) {
	#if ! defined __WXMSW__ && ! defined __WXMAC__
		dynamic[key] = htmlSpecialChars ? DSCgi::HtmlSpecialChars(value) : value;
	#else
		dynamic[key] = value;
	#endif
}

wxString DynSoft::DSTemplate::GetDynamicVariable(const wxString &key) {
	return dynamic[key];
}

wxString DynSoft::DSTemplate::Parse() {
	wxString tmp(content);

	DSStringHashMap::iterator it;
	for(it = variables.begin(); it != variables.end(); ++it)
		tmp.Replace(_N("{s:") + it->first + _N("}"), it->second);
	for(it = intern.begin(); it != intern.end(); ++it)
		tmp.Replace(_N("{i:") + it->first + _N("}"), it->second);
	for(it = dynamic.begin(); it != dynamic.end(); ++it)
		tmp.Replace(_N("{d:") + it->first + _N("}"), it->second);
	parsedContent.Append(tmp);

	return tmp;
}

void DynSoft::DSTemplate::Show(const wxString &header) const {
	const char *tmpHeader = DSCasts::ToChar(header);
	const char *tmpContent = DSCasts::ToChar(parsedContent);
	printf(
		"%s%s\n",
		tmpHeader,
		tmpContent
	);
	delete [] tmpContent;
	delete [] tmpHeader;
}

DynSoft::DSClassInfo DynSoft::DSTemplate::GetClassInfo() {
	return DynSoft::DSClassInfo(
		DS_TEMPLATE_NAME,
		DS_TEMPLATE_VERSION,
		DS_TEMPLATE_AUTHOR
	);
}

void DynSoft::DSTemplate::ClearParsed() {
	parsedContent.Clear();
}

void DynSoft::DSTemplate::SetEntity(DynSoft::DSEntity *entity) {
	if(!entity)
		return;
	
	DynSoft::DSEntity::DSEntityType *entityType = NULL;
	const unsigned int count = entity->GetCount();
	for(unsigned int i = 0; i < count; i++) {
		entityType = entity->Get(i);
		if(entityType)
			SetVariable(entityType->GetKey(), entityType->ToString(), true);
	}
}
